### ISSUE TITLE
Optional chaining control flow cannot narrow discriminated unions.

### ISSUE THREAD

**Issue by whzx5byb** [microsoft/TypeScript#42120]
# Bug Report

### 🔎 Search Terms

optional chaining discriminated union control flow

### 🕗 Version & Regression Information

- This is the behavior in every version I tried.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.1.3#code/JYOwLgpgTgZghgYwgAgAoHsDOwzHSZAbwFgAoZZADwC5kQBXAWwCNoBuMigT1oZfc7IE6RgAd6kACYB+XhABuA0gF8yZUJFiIUAJQgAbOLkUZsufEUE06TVlGQAfZJjBRQAcw7lkPG-3tOLm4gnoLCYhIQkrSmOHggXqqkZJIQCIZQKPJw9qJYMVhxFk56hsYQseYJaqQw9CAIVcjuEGAAwiLiUgAaABR5mAVm8Y7IpUbAJoVVAJS8ttCW3plg9FAEAwB04V1R0puUyNLSyFuUXhQA9JfI0FDo9gAW0BCbyACilKJpYMhg6MhWH8uN9kAAiPh2MEAGkBEmQiFWcH0+i44Mhi0Crg8YMSNWAMGQ-Sw206kRkB2QAEIALw05D1VIwUBRGZLCgDNjIa4fL4-KJ-AFAsAglBg8blSrxGFw36I+jI1HgqXFMYGCZTYb4MGCLY7clcwQ8gCSDXw2BcEAaKGemTeRpuxsJYIGYOQwEwyumIxK6sl3u1sLAzwILpJ+qkbswj3Q9H0kkBYpVBCcjIgzJAUTBmwdyCd4Nd7s94r9kwqAZAMuDVoL4bJkecMbjCaBYOT2aJAEEQAngx6i7c+Y0opsZrmAELw21iwv9tsV0YSsvt2HMeFhzCkiINuftnPeHkAFWe7rNIAtkGtyEYcAA1ihGCg+yBb+7fv24HD3KfLXAE+hCU-TB6HcFoXHifdlCAA)

### 💻 Code

```ts
interface Position {
  x: number;
  y: number;
  computed?: never;
}

interface RelativePosition {
  x: number | string;
  y: number | string;
  computed: Position;
}

declare var pos: Position | RelativePosition;

function getComputedX(pos: Position | RelativePosition): number {
  return pos.computed?.x ?? pos.x;
  // error here. Expect to be type "number", but actually "number | string";
}

if (pos.computed?.x !== undefined) {
  pos; // Expected to be type "RelativePosition", but actually "Position | RelativePosition"
  pos.computed; 
  // Inconsistence here. 
  // If "pos" is "Position | RelativePosition", then "pos.computed" should be "Position | undefined".
  // If "pos" is "RelativePosition", then "pos.computed" should be "Position". (And this is expected.)
  // But here "pos" is "Position | RelativePosition", but "pos.computed" is "Position".
  // The inconsistence make me think it is a bug instead of a suggestion.
}

```

Possibly related: https://github.com/microsoft/TypeScript/pull/34597

**Comment by andrewbranch** [comment]
I think this is essentially a duplicate of #32399/#18758, but is notably not fixed by #38839 (it looks like maybe that’s limited to narrowing with `typeof`).

**Comment by andrewbranch** [comment]
The original repro in #18758 can be [changed very slightly](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDeVSQFxQEQEMtQF8BYAKFLWgA0FlVwJMYAaKHTAZ2ACcBLAOwDmhcvSgBNGigoB+TAFc+AEwgAzfhEUsARh278hxMiWUBjADY4u0MxGBQAHpmoAfcaVI8VUABT2AdDIBovAh2HgAlMikAJD+OFAA9AlQACqiAKJcXAD2XKSGQA) to show this bug:

```ts
type A = { type: "a" }

type X = { type: A, a: string }
type Y = { type?: undefined, b: string }

declare let x: X | Y

if (x.type?.type === "a") {
	x.a // Type Error
}
```

**Comment by ShuiRuTian** [comment]
Interesting. I remember I try to handle expression with optional chain, but it seems I do not do it well. 
However, I would have a look at this.
Thanks a lot for the case @whzx5byb and analytics @andrewbranch !

**Comment by ShuiRuTian** [comment]
This is fixed in #38839 now!
